File: 02-moderate/sudoku/sudoku.py
from sys import argv


def parse_problem(file_path):
    """
        Parse CodeEval each set a time for Sudoku.
    """
    with open(file_path, 'r') as in_file:
        for line in in_file:
            grid, seq =line.split(';')
            grid = int(grid)
            seq = seq.strip().split(',')

            matrix = [seq[i-grid:i] for i in range(grid, (grid*grid) + 1, grid)]
            yield(matrix)


def gfq_sets(matrix, N, size):
    """
        Generate all possible square quadrants of a matrices. Also flattern
        these matrices and only include unique entries.

        Abbriviation = generate flat quadrant sets.
    """
    for y in range(size, N + 1, size):
        y_slice = matrix[y-size:y]
        for x in range(size, N + 1, size):
            yield({i for row in y_slice for i in row[x-size:x]})


def sudoku(matrix):
    """
        Check whether a matrix is a valid sudoku entry.
    """
    N = len(matrix)
    quadrant_size = 3 if N == 9 else 2

    # Check rows
    if any(len(set(row)) < N for row in matrix):
        return(False)

    # Check colums
    elif any(len({matrix[x][y] for y in range(N)}) < N for x in range(N)):
        return(False)

    # Check quadrants
    elif any(len(quad) < N for quad in gfq_sets(matrix, N, quadrant_size)):
        return(False)

    else:
        return(True)


if __name__ == '__main__':
    for problem in parse_problem(argv[1]):
        print(sudoku(problem))

File: 02-moderate/sudoku/sudoku.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <sstream>

using namespace std;

void parse_problem(string line, vector<char> &puzzle, int &N);
bool is_valid_sudoku(vector<char> &puzzle, int N);


int main(int argc, char* argv[]) {
  ifstream in_file(argv[1]);
  string line;

  while (getline(in_file, line)) {
    vector<char> puzzle;
    int N;

    parse_problem(line, puzzle, N);

    if (is_valid_sudoku(puzzle, N))  cout << "True\n";
    else cout << "False\n";
  }

  return(0);
}


void parse_problem(string line, vector<char> &puzzle, int &N) {
  // Parse CodeEval each set a time for Sudoku.
  stringstream sstr(line);
  char num;
  sstr >> N;
  sstr.ignore();
  puzzle.reserve(N*N);

  while (sstr >> num) {
    puzzle.push_back(num);
    sstr.ignore();
  }
}


bool is_valid_sudoku(vector<char> &puzzle, int N) {
  // Check whether a matrix is a valid sudoku entry, only valid for N={3, 9}.
  int quadrant_size;

  if (N == 9) quadrant_size = 3;
  else if (N == 4) quadrant_size = 2;

  vector<int> rowTrack(N, 0), colTrack(N, 0), boxTrack(N, 0);

  for (int i = 0; i < puzzle.size(); i++) {
    int row = i / N;
    int col = i - (row * N);
    int boxRow = row / quadrant_size;
    int boxCol = col / quadrant_size;
    int box = (boxRow*quadrant_size) + boxCol;

    int value = puzzle[i] - '0';

    int mask = 1 << value;

    if ((rowTrack[row] & mask)!=0) return(false);
    if ((colTrack[col] & mask)!=0) return(false);
    if ((boxTrack[box] & mask)!=0) return(false);

    rowTrack[row] += mask;
    colTrack[col] += mask;
    boxTrack[box] += mask;
  }	
  return(true);
}


